util/helpers.py

```diff
@@ -1,8 +1,147 @@
+        If we want to use all layers from the source model except the last one, we use this:
+        drop_source_layers=1
+        '''
+        if drop_source_layers>=6:
+            log_error('The model only has 6 layers, but you are trying to drop '
+                      'all of them or more than all of them. Continuing and '
+                      'dropping only 5 layers')
+        
+        dropped_layers = ['2', '3', 'lstm', '5', '6'][-1 * min(5, int(drop_source_layers)):]
+
+        # Initialize all variables needed for DS, but not loaded from ckpt
+        for v in load_vars:
+            if any(layer in v.op.name for layer in dropped_layers):
+                init_vars.add(v)
+                
+        load_vars -= init_vars
+
+    for v in init_vars:
+        print("Initializing variable from scratch:", v.op.name)
+    for v in load_vars:
+        print("Loading variable from model:", v.op.name)
+        
+    init_op = tfv1.variables_initializer(list(init_vars))
+    for v in list(load_vars):
+        v.load(ckpt.get_tensor(v.op.name), session=session)
+    session.run(init_op)
+
+
+def check_model(checkpoint_filename):
+    checkpoint = tf.train.get_checkpoint_state(FLAGS.checkpoint_dir, checkpoint_filename)
+    if not checkpoint:
+        return False
+    return checkpoint.model_checkpoint_path
+
+
+def try_model(session, load_flag):
+
+    '''
+    if auto, cascasde from last --> best --> init
+  
+    if last, try last, if not found , break
+    
+    '''
+    def try_last():
+        return check_model('checkpoint')
+        
+    def try_best():
+        return check_model('best_dev_checkpoint')
+        
+    def try_init():
+        log_info('Initializing variables...')
+        # Initialize a new model from scratch
+        session.run(tfv1.global_variables_initializer())
+        return True
+    
+    def try_auto():
+        res = try_last()
+        if not res:
+            res = try_best()
+            if not res:
+                res = try_init()
+        return res
+    
+    checkpoint_path = False
+    if load_flag == 'last':
+        # returns false or a checkpoint_path
+        checkpoint_path = try_last()
+        if not checkpoint_path:
+            log_error('Unable to load LAST model from specified checkpoint dir'
+                      ' - consider using load option "auto" or "init".')
+            sys.exit(1)
+            
+    if load_flag == 'best':
+        # returns false or a checkpoint path
+        checkpoint_path = try_best()
+        if not checkpoint_path:
+            log_error('Unable to load BEST model from specified checkpoint dir'
+                      ' - consider using load option "auto" or "init".')
+            sys.exit(1)
+            
+    if load_flag == 'init':
+        # no return
+        try_init()
+        
+    if load_flag == 'auto':
+        # returns true or a checkpoint path
+        checkpoint_path = try_auto()
+
+    if not checkpoint_path in [True,False]:
```

Comment:
nit: Pythonic way is `if checkpoint_path not in [True, False]:`

After you change the `get_checkpoint_path` function to return None instead of False, and `try_init` to return None instead of True, this can just be `if checkpoint_path:` which is much easier to understand.

Reply:
fixed

Reply:
outdated